### PR TITLE
Convert --toolchain_resolution_debug to take a regex of toolchain typ…

### LINKED ISSUE #7713: Minimize output of --toolchain_resolution_debug
### Description of the problem / feature request:

--toolchain_resolution_debug contains useful information for understanding why a toolchain was selected, but it's also very verbose - in an example case involving the Go rules, 1600 lines of output mostly of the form "considering X; constraint doesn't match; rejected X", plus a large number of repeated lines when unspecified actions choose a toolchain (_"Selected execution platform //:builder_platform, type @io_bazel_rules_go//go:toolchain -> toolchain @go_sdk//:go_linux_amd64-impl"_ 800 times)

### Feature requests: what underlying problem are you trying to solve with this feature?

It'd be nice if this could be more compacted, and easier to scan. E.g. by collecting up the resolution details first, then producing a compact structured output second - 

    Toolchain resolution debug info:
        @io_bazel_rules_go//go:toolchain for @bazel_tools//platforms:target_platform
            Matched execution platforms and toolchains: {//:builder_platform -> @go_sdk//:go_linux_amd64-impl, @bazel_tools//platforms:host_platform -> @go_sdk//:go_linux_amd64-impl}
            Rejected toolchains @go_sdk//:go_android_386-impl, @go_sdk//:go_android_amd64-impl, and 20 others for not matching constraint @bazel_tools//platforms:os , value @bazel_tools//platforms:linux
        @io_bazel_rules_go//go:toolchain for @bazel_tools//platforms:host_platform
            Matched execution platforms and toolchains: {//:builder_platform -> @go_sdk//:go_linux_amd64-impl, @bazel_tools//platforms:host_platform -> @go_sdk//:go_linux_amd64-impl}
            Rejected toolchains @go_sdk//:go_android_386-impl, @go_sdk//:go_android_amd64-impl, and 20 others for not matching constraint @bazel_tools//platforms:os , value @bazel_tools//platforms:linux
        @bazel_tools//tools/cpp:toolchain_type for @bazel_tools//platforms:target_platform
        ...

For matching of toolchains against actions themselves, it seems it'd be better to either omit those lines, or to emit the target as well - 800 lines saying the go toolchain was selected but without context don't quite solve that well.

### Bugs: what's the simplest, easiest way to reproduce this bug? Please provide a minimal example if possible.

Add --toolchain_resolution_debug to any build including Go.

### PR DESCRIPTION
**Pull request by katre**
…es to debug

Work towards #12110.
Fixes #7713.

RELNOTES: The flag `--toolchain_resolution_debug` now takes a regex argument, which is used to check which toolchain types should have debug info printed. You may use `.*` as an argument to keep the current behavior of debugging every toolchain type.

### CODE REVIEW

File: src/main/java/com/google/devtools/build/lib/analysis/PlatformConfiguration.java [hunk 1]
```diff
@@ -111,4 +114,23 @@ public Label getTargetPlatform() {
...
+  public boolean debugToolchainResolution(Label toolchainType) {
+    return debugToolchainResolution(ImmutableList.of(toolchainType));
+  }
+
+  /**
+   * Returns true if toolchain resolution debug info should be printed for any of these toolchain
+   * types.
+   */
+  public boolean debugToolchainResolution(Collection<Label> toolchainTypes) {
+    if (toolchainTypes.isEmpty()) {
```
Comment (gregestren) [inline src/main/java/com/google/devtools/build/lib/analysis/PlatformConfiguration.java:128]: If this is empty, is there any actual information to debug? i..e what gets printed when this returns true and the input is empty?

File: src/main/java/com/google/devtools/build/lib/analysis/PlatformConfiguration.java [hunk 2]
```diff
@@ -111,4 +114,23 @@ public Label getTargetPlatform() {
...
+      // Check an empty string, in case the filter is .*
+      return this.toolchainResolutionDebugRegexFilter.test("");
```
Comment (gregestren) [inline src/main/java/com/google/devtools/build/lib/analysis/PlatformConfiguration.java:130]: Does plain old `--toolchain_resolution_debug` still work? (as an alias for `.*`)?

File: src/main/java/com/google/devtools/build/lib/analysis/PlatformConfiguration.java (hunk 1 above)
Comment (katre) [reply src/main/java/com/google/devtools/build/lib/analysis/PlatformConfiguration.java:128]: Even targets with no toolchains will have an execution platform selected (and this is used in some tests).

Ideally we could instead do `--toolchain_resolution_debug=//some/configured:target`, but that would require ConfiguredTargetFunction to make the check and pass a debug flag, thus complicating the ToolchainContextKey even more.

File: src/main/java/com/google/devtools/build/lib/analysis/PlatformConfiguration.java (hunk 2 above)
Comment (katre) [reply src/main/java/com/google/devtools/build/lib/analysis/PlatformConfiguration.java:130]: No, the argument is required (and if it is skipped an error will be printed). I couldn't figure out a way to distinguish between no flag being passed, and a base `--toolchain_resolution_debug` with no argument in order to keep the current behavior unchanged.

**Review by gregestren — APPROVED** [review]

**Pull request closed without merging**